### PR TITLE
Kaito deployment & ux updates

### PR DESCRIPTION
**Pull request by tejhan**
This PR introduces a rehaul for the Kaito deployment integration.

- Added ability to deploy models
- Added tracking of model deployment status
- Added new model deployment menu panel
- Updated model specifications & sources
- Disabled deployment for incompatible models
- Added tooltips for model deployment
- Overall improved UI & accessibility
- Updated CRD generation

image

### CODE REVIEW

File: src/webview-contract/webviewDefinitions/kaito.ts [hunk 1]
```diff
@@ -44,6 +44,17 @@ export type ToWebViewMsgDef = {
         workspace: WorkspaceCRD;
     }; // to webview after workspace CRD is generated
 };
+export type ToWebViewMsgDef2 = {
```
Comment (hsubramanianaks) [inline src/webview-contract/webviewDefinitions/kaito.ts:47]: you can add deploymentProgressUpdate to type ToWebViewMsgDef as well. I am wondering why we need another type ?

File: webview-ui/src/Kaito/state.ts [hunk 2]
```diff
@@ -45,3 +55,36 @@ export const vscode = getWebviewMessageContext<"kaito">({
...
+export const vscode2 = getWebviewMessageContext<"kaitoModels">({
+    generateCRDRequest: null,
+    deployKaitoRequest: null,
+    workspaceExistsRequest: null,
+    updateStateRequest: null,
+    resetStateRequest: null,
+    cancelRequest: null,
+});
+
+export const stateUpdater2: WebviewStateUpdater<"kaitoModels", EventDef, DeploymentState> = {
```
Comment (hsubramanianaks) [inline webview-ui/src/Kaito/state.ts:68]: Can you explain why we would need stateUpdater2 and can't reuse previous stateUpdater ? just for the documentation purpose.

**Review by hsubramanianaks — APPROVED** [review]
Looks good, I had minor comments/suggestions. Thanks @tejhan

File: src/webview-contract/webviewDefinitions/kaito.ts (hunk 1 above)
Comment (tejhan) [reply src/webview-contract/webviewDefinitions/kaito.ts:47]: This is actually outdated code, this has been removed in a further PR 👍

File: webview-ui/src/Kaito/state.ts (hunk 2 above)
Comment (tejhan) [reply webview-ui/src/Kaito/state.ts:68]: Gotcha, I'll make sure to add documentation for this in further PR 👍

Comment (hsubramanianaks) [reply webview-ui/src/Kaito/state.ts:68]: I hope this has been refactored in future PR's ?

File: src/panels/KaitoModelsPanel.ts [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        });
+    }
+    convertAgeToMinutes(creationTimestamp: string): number {
+        const createdTime = new Date(creationTimestamp);
+        const currentTime = new Date();
+        const differenceInMilliseconds = currentTime.getTime() - createdTime.getTime();
+        const differenceInMinutes = Math.floor(differenceInMilliseconds / 1000 / 60);
+        return differenceInMinutes;
+    }
+    statusToBoolean(status: string): boolean {
```
Comment (Tatsinnit) [inline src/panels/KaitoModelsPanel.ts:156]: 💡 https://github.com/Azure/vscode-aks-tools/pull/963#pullrequestreview-2365020283

also some recommendations here: https://github.com/Azure/vscode-aks-tools/pull/963#discussion_r1792833650

File: src/panels/KaitoPanel.ts [hunk 4]
```diff
@@ -117,12 +119,14 @@ export class KaitoPanelDataProvider implements PanelDataProvider<"kaito"> {
...
-        webview.postGetWorkspaceResponse({
-            workspace: {
-                workspace: "workspace CRD yaml",
-            },
-        });
+        // webview.postGetWorkspaceResponse({
+        //     workspace: {
+        //         workspace: "workspace CRD yaml",
+        //     },
+        // });
```
Comment (Tatsinnit) [inline src/panels/KaitoPanel.ts:126]: 💡Observation: Do we need these commented line here? if not lets remove any unwanted commented code please? = )

File: src/panels/KaitoPanel.ts [hunk 5]
```diff
@@ -190,6 +174,51 @@ export class KaitoPanelDataProvider implements PanelDataProvider<"kaito"> {
             return this.resourceManagementClient.resources.getById(this.armId, "2023-08-01");
         });
 
+        // Install kaito enablement
+        const kaitoInstallationResult = await longRunning(`Enabling the kaito for this cluster.`, () => {
```
Comment (Tatsinnit) [inline src/panels/KaitoPanel.ts:178]: 💡 idea: shall we add `${clusterName}` here to be more explicit, just some thoughts?

File: src/panels/KaitoPanel.ts [hunk 6]
```diff
@@ -259,61 +291,74 @@ export class KaitoPanelDataProvider implements PanelDataProvider<"kaito"> {
...
-                }
-                await this.installKaitoFederatedCredentials(
-                    clusterInfo.result.nodeResourceGroup!,
-                    this.clusterName,
-                    aksOidcIssuerUrl,
-                );
+        const aksOidcIssuerUrl = clusterInfo.result.oidcIssuerProfile?.issuerURL;
+        if (!aksOidcIssuerUrl) {
+            vscode.window.showErrorMessage(`Error getting aks oidc issuer url`);
+            return { succeeded: false, error: "Error getting aks oidc issuer url" };
```
Comment (Tatsinnit) [inline src/panels/KaitoPanel.ts:322]: 💡 Thoughts only: Is there any form of `err` above code returns, if yes we can add that in the end of the `error` as extra information if that helps, please.

**Review by Tatsinnit — APPROVED** [review]
❤️ Thank you so much @tejhan!! Looks nice! I have shared just few key observations in this PR, please remove any commented code you think we don't need in release.

For wider folks and for archival reason, this code has less comments in review because this PR is being reviewed previously multiple times, so we are not giving any hallway pass but just the nature of this PR relevance and as a historic reason, adding bunch of PRs where we did take a look at this. 

* Here was the last reviewed comment and PR for it: https://github.com/Azure/vscode-aks-tools/pull/963#pullrequestreview-2365020283 
* If this PR is already cherry-picked please close this PR: https://github.com/tejhan/vscode-aks-tools/pull/1
* Some extraction recommendation here: https://github.com/Azure/vscode-aks-tools/pull/963#discussion_r1798619283
* Detail here is just for some thought share and we can do this in separate PR please: https://github.com/Azure/vscode-aks-tools/pull/963#discussion_r1798619120

Looks good to me, lets definitely get this tested with folks and fyi to @qpetraroia  for six shakedown and we can merge this, just few comments from my side in this PR

File: src/panels/KaitoPanel.ts (hunk 4 above)
Comment (hsubramanianaks) [reply src/panels/KaitoPanel.ts:126]: this is taken care in this PR - https://github.com/tejhan/vscode-aks-tools/pull/2/files

File: src/panels/KaitoPanel.ts (hunk 6 above)
Comment (hsubramanianaks) [reply src/panels/KaitoPanel.ts:322]: we already handle it in above lines where we retrieve clusterInfo from getManagedClusterCall, this one specifically if the issuerURL is undefined/empty/null after GET managed cluster call - we can finetune the error message to be specific if that's the case 
https://github.com/Azure/vscode-aks-tools/blob/4760b21418889481e7268f23f8a781eabb5c9abf/src/panels/KaitoPanel.ts#L307

I have improved the error message in this PR - https://github.com/tejhan/vscode-aks-tools/pull/2/files

**Comment by Tatsinnit** [comment]
> ❤️ Thank you so much @tejhan!! Looks nice! I have shared just few key observations in this PR, please remove any commented code you think we don't need in release.
> 
> For wider folks and for archival reason, this code has less comments in review because this PR is being reviewed previously multiple times, so we are not giving any hallway pass but just the nature of this PR relevance and as a historic reason, adding bunch of PRs where we did take a look at this.
> 
> * Here was the last reviewed comment and PR for it: [Kaito Deployment & UX Updates #963 (review)](https://github.com/Azure/vscode-aks-tools/pull/963#pullrequestreview-2365020283)
> * If this PR is already cherry-picked please close this PR: [Refactor to smaller functs. tejhan/vscode-aks-tools#1](https://github.com/tejhan/vscode-aks-tools/pull/1)
> * Some extraction recommendation here: [Kaito Deployment & UX Updates #963 (comment)](https://github.com/Azure/vscode-aks-tools/pull/963#discussion_r1798619283)
> * Detail here is just for some thought share and we can do this in separate PR please: [Kaito Deployment & UX Updates #963 (comment)](https://github.com/Azure/vscode-aks-tools/pull/963#discussion_r1798619120)
> 
> Looks good to me, lets definitely get this tested with folks and fyi to @qpetraroia for vsix shakedown and we can merge this, just few comments from my side in this PR
> 
>  

Thank you so much, @tejhan  ! We had a quick chat and a demo today, and it seems that most of the points mentioned have already been covered and reviewed in this PR. I'll go ahead and merge it, and we can address any remaining comments in the next PR aka https://github.com/Azure/vscode-aks-tools/pull/1003 

I really appreciate your time and feedback! Thanks again! fyi to @hsubramanianaks + @qpetraroia ❤️

File: webview-ui/src/Kaito/state.ts (hunk 2 above)
Comment (Tatsinnit) [reply webview-ui/src/Kaito/state.ts:68]: 📓 Nice, seems like @tejhan  mentioned that he will take care of most of this in PR here post this merge: https://github.com/Azure/vscode-aks-tools/pull/1003 , leaving this as fyi comment.

File: src/panels/KaitoModelsPanel.ts (hunk 3 above)
Comment (Tatsinnit) [reply src/panels/KaitoModelsPanel.ts:156]: 📓 Nice, seems like @tejhan mentioned that he will take care of most of this in PR here post this merge: https://github.com/Azure/vscode-aks-tools/pull/1003 , leaving this as fyi comment.

**Pull request merged**